feat: support jumping from non-file-based buffers (e.g. org capture)

Hugo-Heagren opened the pull request:
The package doesn't work for non-file based buffers, which in particular means it doesn't work in org (roam) capture buffers (for jumping to the pdf I'm making notes on).

I appreciate that you need some sort of reference to get off the ground, and that *normally* that will be a file name, but for my use case it would be great to be able to use other things as well (org roam node id, or a citation ref). Perhaps each jumper could have a parameter which gets the value to use as the jumping "input" or "start". If not present, the paramater could just default to the function `buffer-file-name`.

This would make a big difference to me and my workflows.

Hugo-Heagren commented:
Just pushed some new commits -- I've dealt with place formatting.

I've also slightly reworked how methods are selected for jumpers. I'll explain why with an example.

I was testing my changes, and I found errors when `related-files-jumpers` included any recipe jumpers. This was because all places are called with all jumpers, and there was no restriction on what kind of place a recipe jumper could take---it would be called with something like `#<buffer scratch>`, and then error when it tried to do string manipulation on that. 

I could see this being a problem in general for users with more than one kind of jumper, hence the changes.

The other two things above still stand. I also need to make the interface for *making* places generic as well.

DamienCassou commented:
> I could implement this sort of behaviour [multi-category] if you would like

this makes sense in my opinion

> Compared with the elegance and automatic type-checking of the generics/methods throughout the rest of the code, this is kind of ugly

`pcase` could make it a little better

> One (rather laborious) way of handling this better would be to define another generic...

I think we can keep this problem for later

DamienCassou commented on related-files-recipe.el:
```diff
@@ -36,7 +36,7 @@
 
 ;;; Overrides of Public Methods
 
-(cl-defmethod related-files-apply ((jumper (head recipe)) place)
```
this is great! This kind of jumpers only apply to filename places so this change makes perfect sense.

Hugo-Heagren replied:
Nice, thanks :)

DamienCassou commented on related-files.el:
```diff
@@ -234,35 +247,71 @@ ones.  Look at `related-files-jump' and `related-files-make' if you don't
+The default implementation assumes PLACE is a filename."
+  (file-exists-p place))
+
+(cl-defmethod related-files-place-exists-p ((place buffer))
```
why is there no implementation of this method for filenames?

Hugo-Heagren replied:
When I wrote this, I was still following the earlier practice of writing a defgeneric with non-specialised arguments, and with a default iimplementation which assumed the place was a filename (so the implementation for filenames is just the default implementation).

Do you think it would be better to completely separate the definitions for all these generics/methods (`place-exists-p` and`goto-place`

DamienCassou replied:
I've no strong feeling about this so pick whatever you want. It feels to me that the default for `related-files-place-exists-p` should probably return `nil` and the default for `related-files-goto-place` should throw.

DamienCassou commented on related-files.el:
```diff
@@ -234,35 +247,71 @@ ones.  Look at `related-files-jump' and `related-files-make' if you don't
+The default implementation assumes PLACE is a filename."
+  (find-file place))
+
+(cl-defgeneric related-files-goto-place ((place buffer))
```
why is there no implementation of this method for filenames?

Hugo-Heagren replied:
Same as above:



When I wrote this, I was still following the earlier practice of writing a defgeneric with non-specialised arguments, and with a default iimplementation which assumed the place was a filename (so the implementation for filenames is just the default implementation).

Do you think it would be better to completely separate the definitions related-files-place-exists-p and realted-files-goto-place and have the filename methods defined separately? It might be cleaner.

DamienCassou commented on related-files.el:
```diff
@@ -234,35 +247,71 @@ ones.  Look at `related-files-jump' and `related-files-make' if you don't
+  "Call `switch-to-buffer'."
+  (switch-to-buffer place))
+
+(cl-defgeneric related-files-attach-jumper-to-place (jumper place)
```
why do you need a method for this? Do you plan to override it?

Hugo-Heagren replied:
> Do you plan to override it?

Yes. because different places will need to have extra information (the jumper) 'attached' to them in different ways -- only filenames (strings) can be propertized (this was what we using before). The implementation in my init file of my org-roam jumper returns an org roam node as a place. This is a cl struct object, so can't be propertized, so I needed to write a different way of attaching something to it.

(I haven't got there yet, but soon I'll also write a corresponding method to *retrieve* the jumper *from* the place. I haven't got to generlising making new places yet though.)

DamienCassou reviewed (changes requested):
I've only done a quick review to give early feedback. Feel free to ask me for another review when you feel ready for it.

DamienCassou commented:
This is excellent work @Hugo-Heagren! Thank you very much.

DamienCassou commented:
Thank you for your answers. You are on the right track I think.

Hugo-Heagren commented:
I think the bulk of the work is now finished. I had a play around with what's currently on the repo today, and I was able to setup my workflow for capturing to org-roam nodes from pdfs exactly as I wanted it. I haven't set up my other jumpers yet, but I am confident the system as it stands is flexible enough.

I think there are two more things I need to do:
- modify the tests to align with the way the new API works  (or find out why else the tests are failing!)
- Decide what to do about the default implementation thing discussed above, and implement that decision across the code

A final tiny thing. I think it might be worth adding an optional argument to `related-files-format-place` to indicate whether a place is to be formatted as if it currently exists or not. At the moment this is done by `related-files--format-place` (by appending `(create it!)` if the place does not exist). The problem with this approach is that it might not be appropriate for some styles of formatting. My org-roam node formatted prints my nodes with tags at the end (e.g. `Title of Node <alignment space> #list #of #tags #here`), and on my laptop that can fill up the width of the screen, so appended text wouldn't be the best way of displaying these.

Since there's already an optional argument for that method (`annotate`), the best approach might be to make them both into keyword arguments.

What do you think?

DamienCassou commented:
I haven't looked at the code. If you feel ready for a review, please tell me.

Regarding the formatting code, I'm not quite sure how you plan to use `annotate`. From the code it seems this argument is always `nil` anyway.

I'm not sure the places that are listed by `related-files--choose-place` should have different ways to say that they need to be created first. It would feel weird that some places show "(create it!)" while others show something completely different (e.g., a different color).

We could try something that works in all the cases we have for now. For example a different color (e.g., a face inherting `font-lock-doc-face`) or an initial `+` character.

Hugo-Heagren commented:
> Regarding the formatting code, I'm not quite sure how you plan to use `annotate`. From the code it seems this argument is always `nil` anyway.

It was introduced because it is sometimes important (i.e. in passing targets to embark) that the string returned does not have any extra information in it. It is non-nil in the call to `related-files--completing-read` inside `related-files--choose-place`. See the new docstring for `related-files--completing-read`, which specifies that:

> FORMATTER is a function for formating ENTITIES. It should return a string for presentation to the user. It takes one entity, and one optional argument ANNOTATE. If ANNOTATE is non-nil, it is acceptable (though not necessary) to return a string with annotations or 'decorations'. It ANNOTATE is nil, only a 'bare' version should be returned.

I made the argument optional so that people writing their own setups who weren't fussed about supporting embark could just ignore it. See here:

https://github.com/Hugo-Heagren/related-files/blob/32ff43d559cceacc625c548db04ea6eeb7820413/related-files.el#L543

> I'm not sure the places that are listed by `related-files--choose-place` should have different ways to say that they need to be created first. It would feel weird that some places show "(create it!)" while others show something completely different (e.g., a different color).
> 
> We could try something that works in all the cases we have for now. For example a different color (e.g., a face inherting `font-lock-doc-face`) or an initial `+` character.

Well I suggested this and personally I still think it's a good idea, especially for candidates with quite rich formatting. But it's your package and everything else works pretty much perfectly for me, so I'm happy to agree to disagree and leave it for now?

DamienCassou commented:
> leave it for now?

agree. Let's keep it simple for now. The PR is already huge.

Hugo-Heagren commented:
### Testing
I looked at the tests today. Many of them fail. A major reason for this is that they rely on jumpers defined as functions, by let-binding a variable name to a lambda. The tests then fail because `related-files-apply` has a method defined for *symbols* (which includes normal function), but this doesn't include lambdas. There is no specialiser for functions in general, so the lambda-definitions are handled by the default implementation `related-files-apply`, which just returns nil.

This isn't really a problem in the application itself, becuase I don't think people are very likely to use lambdas in `related-files-jumpers`, but it's making the tests fails.

How would you like to handle this? Since a lot of the lambdas are quite similar, we could just define a few jumpers at the top of the file with `defun`, and then reuse them all throughout the tests?

### Organisation
I finished the implementation of the methods for buffers, and separated everything properly. There's now a lot of code in the section called `Jumpers and Places Public API`, a lot of which is specific implementations. I think it might be a good idea to move this into separate files (`related-files-buffer`, `related-files-file`) in the same way that the recipe code is separate from the generic definitions.

What  you think?

Apart from these things, this PR is ready for review.

DamienCassou commented:
> How would you like to handle this? Since a lot of the lambdas are quite similar, we could just define a few jumpers at the top of the file with defun, and then reuse them all throughout the tests?

this looks reasonable to me

> I think it might be a good idea to move this into separate files (related-files-buffer, related-files-file) in the same way that the recipe code is separate from the generic definitions.

excellent idea!

> Apart from these things, this PR is ready for review.

If that is ok with you, I will review when the code is moved to separate files.

Hugo-Heagren commented:
I've moved the code into separate files, and updated the tests. Al of the tests pass on my machine, but they fail on the github CI. I'm not sure what's wrong -- could you have a look?

DamienCassou commented:
> I'm not sure what's wrong -- could you have a look?

The exact same tests fail for me as well (Emacs 28.2). Maybe you need to delete the .elc files? Or try to reproduce in `emacs -Q`?

DamienCassou commented on related-files.el:
```diff
@@ -277,44 +333,76 @@ Beyond the filler, this function is called with the :jumper and
+     ,@related-files-place-category-tests
+     (t (type-of ,place))))
+
+(defun related-files--format-place (initial-places place &optional annotate)
```
instead of having the `annotate` boolean, I would probably split this function into 2: one that never add "(create it!)" and one that does when the place doesn't exist.

Hugo-Heagren replied:
I introduced the optional `annotate` arg so that a single formatting function could be passed to `related-files--completing-read`, which would handle both annotated and non-annotated output -- this is important for compatibility with embark (see the `format-function` let-var in `related-files--completing-read`). Separating the function in two would mean this no longer worked, and would make the code inside that function much more complex. Personally I don't think that's worth it.

DamienCassou replied:
very good! Thanks you for explaining.

Hugo-Heagren commented:
> The exact same tests fail for me as well (Emacs 28.2). Maybe you need to delete the .elc files? Or try to reproduce in `emacs -Q`?

Sure, I'll have a look

DamienCassou commented on related-files-buffer.el:
```diff
@@ -0,0 +1,56 @@
+;;; related-files-buffer.el --- Provide methods and jumpers for buffers  -*- lexical-binding: t; -*-
```
I don't think there are any "jumpers" provided in this file. Also, the notion of "place" could probably be added to the documentation. I would write something like:

"Add support for buffer-based places to related-files"

Hugo-Heagren replied:
Fixed in 7b56215

DamienCassou commented on related-files-file.el:
```diff
@@ -0,0 +1,58 @@
+;;; related-files-file.el --- Provide methods and jumpers for files  -*- lexical-binding: t; -*-
```
"Add support for file-based places to related-files"?

Hugo-Heagren replied:
Fixed in 6f2f029

DamienCassou commented on tests/related-files-test.el:
```diff
@@ -38,15 +38,41 @@
+(defvar related-files-test-jumper-args nil)
+
+(defun related-files-test-push-place-jumper (place)
+  (push place related-files-test-jumper-args)
```
I would prefer avoiding tests with side effects. We might quickly get into trouble with tests only passing when executed in a certain order. What about defining this jumper with `cl-labels` or similar?

Hugo-Heagren replied:
> I would prefer avoiding tests with side effects.

I understand, and would usually agree. Strangely though, I've tried  `cl-labels` and `cl-flet`, each nested in a variety of ways in the test, and I just can't make it work. I'd welcome if you could, but at the moment I can't see another way of achieving this without side effects.

DamienCassou replied:
Ok. Thanks for trying.

DamienCassou reviewed (changes requested):
This is brilliant work, thank you so much! I left some small feedback but it feels to me this work is very close to being done.

Great job!!

Hugo-Heagren commented:
> This is brilliant work, thank you so much! I left some small feedback but it feels to me this work is very close to being done.

Thanks for the feedback, I'll have a look at the small reviews now.

The reason the tests are failing is the newly separate file `related-files-file.el` is not loaded, so the (again, now separate) methods for handling files are not available. I can put a line `(require 'related-files-file)` at the top of the test doc, but the tests don't still don't pass in a bar emacs environment until I visit `related-files-file` and eval the buffer.

I'm guessing this might have something to do with how the testing infrastructure is written. I don't know `make` at all well, so you might have to give me some guidance here?

DamienCassou commented:
> I can put a line (require 'related-files-file) at the top of the test doc, but the tests don't still don't pass in a bar emacs environment until I visit related-files-file and eval the buffer.

indeed. In an interactive environment, you have to use eval-buffer to
load the content of related-files-file. In a non-interactive
environment, adding the line `(require 'related-files-file)` to the test
file will do the trick.

I suggest adding the line and pushing the commit: I think the CI will make some progress and run the tests. You can also test locally in a terminal: just type `make test` to run the ERT tests.

When the tests pass, you will have to make the linter happy: `make lint`.

You will find more information about the make targets you can use in https://gitea.petton.fr/DamienCassou/makel/.

Optional: The best thing would be to have a test file for each non-test file:

- related-files-test.el to test the content of related-files.el
- related-files-recipe-test.el to test the content of related-files-recipe.el
- related-files-file-test.el to test the content of related-files-file.el
- related-files-buffer-test.el to test the content of related-files-buffer.el

DamienCassou commented:
Is there anything I can help you with?

Hugo-Heagren commented:
> Is there anything I can help you with?

No it's fine! Sorry, your feedback above was great I just haven't had much time to work on this in the last few days -- I should be able to get back to it this weekend at the latest.

Hugo-Heagren commented on related-files.el:
```diff
@@ -277,44 +333,76 @@ Beyond the filler, this function is called with the :jumper and
+take the relevant form. The return value of each clause must be a
+symbol for a category. See info node `(elisp)Programmed
+Completion' for more information."
+  ;; TODO pcase arg list type? (for :type)
```
Fixed in d5d2941 (new version, edited and force-pushed).

Hugo-Heagren commented on related-files.el:
```diff
@@ -277,44 +333,76 @@ Beyond the filler, this function is called with the :jumper and
     (related-files--make-place place)))
 
-(defun related-files--format-place (initial-directory place)
+(defcustom related-files-place-category-tests
```
Fixed in d5d2941 (new version, edited and force-pushed).

Hugo-Heagren commented on related-files.el:
```diff
@@ -159,12 +159,24 @@ MANUAL (TODO)."
-(defun related-files-jump (&optional jumpers current-place)
-  "Let the user choose where to go from CURRENT-PLACE by asking JUMPERS.
+(defun related-files-jump (&optional jumpers current-places)
+  "Let the user choose where to go from the current place by asking JUMPERS.
```
Understood. This particular one was fixed in 9f066bf865e614ced61b6ca0aff2c755bc0c83fe.

Hugo-Heagren commented on related-files.el:
```diff
@@ -159,12 +159,24 @@ MANUAL (TODO)."
+
+Each function should return the current place in a different
+form, or nil if called in circumstances such that the format does
+not apply (i.e. the function `buffer-file-name' returns nil when
```
Fixed in 314123750fa1e14c944442f30585c3d7130fa1b1 (edited and force-pushed).

Hugo-Heagren commented on related-files.el:
```diff
@@ -159,12 +159,24 @@ MANUAL (TODO)."
+  "List of functions returning the current place.
+
+Each function should return the current place in a different
+form, or nil if called in circumstances such that the format does
```
Fixed in 314123750fa1e14c944442f30585c3d7130fa1b1 (edited and force-pushed).

Hugo-Heagren commented on related-files-buffer.el:
```diff
@@ -0,0 +1,56 @@
+
+;;; Commentary:
+
+;; TODO
```
Fixed in c2208401deaee25d74baa7ec1295c9c4bfb63ccd.

DamienCassou commented:
@Hugo-Heagren are you done and waiting for feedback or do you still have stuff to do?

Hugo-Heagren commented:
> are you done and waiting for feedback or do you still have stuff to do?

Just pushed some new work. I've dealt with all the small reviews above except the two I've left comments on. I need some direction on what to about those two, but other than that I think I'm do. The linter is happy with what I have at the moment.

DamienCassou commented:
We should merge that now, you have been working on it for way too long. I still see 2 problems though:

- `make lint` reports
    ```
    related-files.el:513:1: Error: Symbol’s value as variable is void: related-files-place-category-tests
    ```
- `make test` reports an error with `related-files-test-collect-existing-places-avoids-calling-same-jumper-with-current-place-twice`

Do you want to fix these two problems or should I?

Also, because the commits are not atomic (checks don't pass on each commit), I plan to squash all your work inside just 1 big commit.

Hugo-Heagren commented:
> We should merge that now, you have been working on it for way too long.

Thank you! I don't mind working on big projects for a long time though, and I think was really worth it. I'm pleased with it.

> I still see 2 problems though

I've looked at the first problem and I don't know how to solve it, so it would be great if you could do that. The second problem is because of the stateful test I wrote (remember you left some fedback saying you tried not to use tests which change global state). I removed the commits for it and tried to write a version without global state (i.e. by defining the jumper with `cl-labels` or something), but I really tried and I couldn't make it work. If you can, it would be great to solve that too.

> Also, because the commits are not atomic (checks don't pass on each commit), I plan to squash all your work inside just 1 big commit.

Sure, that sounds good. I haven't heard of 'atomic' commits in this sense before (I have always tried to make commits which are atomic in the sense of making only one change at a time) -- is this a widely used idea?

Hugo-Heagren commented:
> Also, because the commits are not atomic (checks don't pass on each commit), I plan to squash all your work inside just 1 big commit.

Having set this PR to 'ready to merge', I see that I need to sign my commits. How would you like to handle this? I could squash them all at my end, then sign and (force) push that commit to this branch. Or I could retroactively sign all the previous commits (see e.g. [this](https://stackoverflow.com/questions/41882919/is-there-a-way-to-gpg-sign-all-previous-commits) discussion), and push them all signed. Or I could let you handle this whole thing?

DamienCassou commented:
I will try to fix the issues myself.

> I haven't heard of 'atomic' commits in this sense before -- is this a widely used idea?

I don't want to answer this question because too many developers will just say that what they do is best practice and that everyone not doing that is just stupid (I have observed that bad developers tend to do that more than good ones). I can only tell you why I think a commit can only be said to be atomic if it passes the checks (the linter should be happy, the code should conmpile, the tests should pass, the application should work, …).

My opinion is that decomposing into small commits is a good thing as it makes each commit (1) easier to understand and (2) easier to revert. The smaller the better and I frequently create commits modifying just one line (I sometimes even create two commits changing the same line if there are two reasons to change it). Still, I believe that a commit that breaks checks loses its advantages because it strongly depends on the commits fixing the broken checks: (1) you can no longer understand what the commit does separately (as you need to also look at the commits fixing the checks to fully understand) and (2) you can't revert this commit without also finding and reverting the commits fixing the checks.

In my company, there are cases when we break this rule but this is not common. A use case for breaking this rule is when we need to both rename a file and change its content to make all checks to pass. We split that into 2 commits (the rename and then the change) to make sure Git realizes the first commit is a rename operation instead of a deletion of a file and the addition of another unrelated file.

DamienCassou commented:
In the case of this PR, it would be good to rewrite the history into several commits where each one validates all the tests. I don't have the energy nor time to do that on your work though. That's why I suggested to squash everything together which I think is a bit better than having many commits that break the checks. What do you think?

DamienCassou commented:
According to the linter, the use of `related-files-place-category-tests` within the macro `related-files--get-place-category` is going to be problematic. Intuitively, functions depending on the macro might be compiled before the user gets a chance to change the value of `related-files-place-category-tests`. I asked the [question to emacs-devel](https://lists.gnu.org/archive/html/emacs-devel/2023-02/msg00939.html) to be really sure.

Hugo-Heagren commented:
> I suggested to squash everything together ... What do you think?

I understand, and I think this is a good idea. I'll wait until the thing about `related-files-place-category-tests` is resolved, and then squash, sign and push.

> According to the linter, the use of `related-files-place-category-tests` within the macro `related-files--get-place-category` is going to be problematic.

I don't enough about lisp to fully appreciate this problem, or the discussion at emacs-devel, but it seems to me that the simplest way is to rewrite the macro as a normal function. I used a macro originally because I thought the resulting code would be clearer, but having now written a function I think the new code is as clear, if not more. All the same tests still pass (that is, all the tests except the stateful one which you said you'd deal with).

Hugo-Heagren commented:
I've pushed a new, squashed, signed commit.

DamienCassou commented:
I'm very sorry for answering you so late. I thought I would have more time at the beginning of this year but I don't.

Your most recent changes don't work for me. Would you mind deleting related-files' .elc files, restarting Emacs and checking if related-files still works? I'm particularly interested in the completion interface that appears when there are multiple places to go to.

Here is the error I always get:

```
Debugger entered--Lisp error: (invalid-function pcase)
```

Here is the code that is problematic I think:
```emacs
(defun related-files--get-place-category (place)
  "Get completion category for PLACE.

Use `pcase' to check PLACE against the patterns in
`related-files-place-category-tests'.  If none succeed, return
`type-of' PLACE."
  (apply #'pcase place
	 `(,@related-files-place-category-tests
	   (_ (type-of ,place)))))
```

In Emacs 28.2, `pcase` is a macro, not a function. As such, it can't be passed as argument to `apply` which takes a function as argument.

Hugo-Heagren commented:
> In Emacs 28.2, `pcase` is a macro, not a function. As such, it can't be passed as argument to `apply` which takes a function as argument.

Yes, you're right. That code was problematic. I solved the issue by using `eval` instead, which behaves similarly to apply in this case, but also supports macros. It works for me now.

DamienCassou commented:
Thank you very much for your awesome work @Hugo-Heagren! I just had to fix one test to make everything work.
